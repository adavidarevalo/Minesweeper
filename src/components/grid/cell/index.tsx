import { CellState, Coords } from '../../../utils/field';
import { Box, Button } from '@chakra-ui/react'
import Lottie from 'react-lottie';
import bombAnimation from "../../../../lotties/bomb.json"
import flagAnimation from '../../../../lotties/flag.json';
import { useState } from 'react';

interface CellProps {
  children: number;
  coords: Coords;
  onClick: (coords: Coords) => void;
  onContextMenu: (coords: Coords) => void;
}

export default function Cell({ children, onClick, coords }: CellProps) {
  const [showBtnContent, setShowBtnContent] = useState(false);

  const handleClick = () => {
    setShowBtnContent(true);
    onClick(coords);
  };

  return (
    <Button p={'6px'} onClick={handleClick} isDisabled={showBtnContent}>
      {showBtnContent && children === CellState.bomb && (
        <Lottie
          options={{
            loop: true,
            autoplay: true,
            animationData: bombAnimation,
          }}
        />
      )}
      {showBtnContent && children === CellState.mark && (
        <Lottie
          options={{
            loop: true,
            autoplay: true,
            animationData: flagAnimation,
          }}
        />
      )}
      {showBtnContent && children === CellState.empty && (
        <Box h="full" w="full"></Box>
      )}
      {showBtnContent && children === CellState.weakFlag && (
        <Box h="full" w="full"></Box>
      )}
    </Button>
  );
}


