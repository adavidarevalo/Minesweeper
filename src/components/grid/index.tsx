import { Coords, Field } from "@/utils/field";
import {Grid as GridChackra, GridItem} from '@chakra-ui/react';
import Cell from "./cell";

interface GridProps {
  children: Field;
  maxWidth?: string;
  onClick: (coords: Coords) => void;
  onContextMenu: (coords: Coords) => void;
}

export default function Grid({ children, maxWidth = "300px", ...rest }: GridProps) {
  return (
    <GridChackra
      templateColumns={`repeat(${children.length}, 1fr)`}
      gap={2}
      maxW={maxWidth}
    >
      {children.map((row, y) =>
        row.map((cell, x) => (
          <GridItem key={`${y}_${x}_${cell}`}>
            <Cell coords={[y, x]} {...rest}>
              {cell}
            </Cell>
          </GridItem>
        )),
      )}
    </GridChackra>
  );
}
