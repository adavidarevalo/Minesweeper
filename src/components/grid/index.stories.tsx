import { Meta, StoryObj } from '@storybook/react';
import Grid from "./index"
import { MockFieldData } from './grid.mock';

export default {
  title: 'Grid/Table',
  component: Grid,
} as Meta;

type Story = StoryObj<typeof Grid>;

export const GridDefault: Story = {
  args: {
    children: MockFieldData,
    maxWidth: "300px"
  },
};
