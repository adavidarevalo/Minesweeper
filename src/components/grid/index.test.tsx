import { render } from '@testing-library/react';
import Grid from '.';
import { MockFieldData } from './grid.mock';

it('Grid renders correctly', () => {
  const props = {
    onClick: jest.fn(),
    onContextMenu: jest.fn(),
  };

  const { asFragment } = render(<Grid {...props}>{MockFieldData}</Grid>);

  expect(asFragment()).toMatchSnapshot();
});
