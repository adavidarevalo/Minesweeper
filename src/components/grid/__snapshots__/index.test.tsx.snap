// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-f15krj"
  >
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
    <div
      class="css-1gvp5h0"
    >
      <button
        class="chakra-button css-mt4qtn"
        type="button"
      />
    </div>
  </div>
</DocumentFragment>
`;
